pkg/utils/host/host.go

```diff
@@ -20,54 +20,408 @@
+	isHostPidNs, err = isHostNamespace("pid")
+	if err != nil {
+		return err
+	}
+	isHostNetNs, err = isHostNamespace("net")
+	if err != nil {
+		return err
+	}
+
+	initDone = true
+	return nil
+}
+
+// IsHostPidNs returns true if the current process is running in the host PID namespace
+func IsHostPidNs() bool {
+	if !initDone {
+		panic("host.Init() not called")
+	}
+	return isHostPidNs
+}
+
+// IsHostNetNs returns true if the current process is running in the host network namespace
+func IsHostNetNs() bool {
+	if !initDone {
+		panic("host.Init() not called")
+	}
+	return isHostNetNs
+}
+
+// isHostNamespace checks if the current process is running in the specified host namespace
+func isHostNamespace(nsKind string) (bool, error) {
 	selfFileInfo, err := os.Stat("/proc/self/ns/" + nsKind)
 	if err != nil {
-		return false
+		return false, err
 	}
 	selfStat, ok := selfFileInfo.Sys().(*syscall.Stat_t)
 	if !ok {
-		return false
+		return false, fmt.Errorf("reading inode of /proc/self/ns/%s", nsKind)
 	}
 
 	systemdFileInfo, err := os.Stat(fmt.Sprintf("%s/1/ns/%s", HostProcFs, nsKind))
 	if err != nil {
-		return false
+		return false, err
 	}
 	systemdStat, ok := systemdFileInfo.Sys().(*syscall.Stat_t)
 	if !ok {
-		return false
+		return false, fmt.Errorf("reading inode of %s/1/ns/%s", HostProcFs, nsKind)
 	}
 
-	return selfStat.Ino == systemdStat.Ino
+	return selfStat.Ino == systemdStat.Ino, nil
 }
 
-func init() {
-	// Initialize HostRoot and HostProcFs
-	HostRoot = os.Getenv("HOST_ROOT")
-	if HostRoot == "" {
-		HostRoot = "/"
+func AddAutoSdUnitRestartFlag(command *cobra.Command) {
+	command.PersistentFlags().BoolVarP(
+		&autoSdUnitRestartFlag,
+		"auto-sd-unit-restart",
+		"",
+		false,
+		"Automatically run in a privileged systemd unit if lacking enough capabilities",
+	)
+}
+
+// autoSdUnitRestart will automatically restart the process in a privileged
+// systemd unit if the current process does not have enough capabilities.
+func autoSdUnitRestart() (exit bool, err error) {
+	const IgInSystemdUnitEnv = "IG_IN_SYSTEMD_UNIT"
+
+	// No recursive restarts
+	if os.Getenv(IgInSystemdUnitEnv) == "1" {
+		return false, nil
+	}
+
+	// Only root can talk to the systemd socket
+	if os.Geteuid() != 0 {
+		return false, nil
+	}
+
+	// This workaround is meant for the "kubectl debug node" node. For now,
+	// don't attempt it if we are obviously not running in that context.
+	if HostRoot != "/host" || os.Getenv("KUBERNETES_PORT") == "" {
+		return false, nil
+	}
+
+	// If we already have CAP_SYS_ADMIN, we don't need a workaround
+	c, err := capability.NewPid2(0)
+	if err != nil {
+		return false, err
+	}
+	err = c.Load()
+	if err != nil {
+		return false, err
+	}
+	if c.Get(capability.EFFECTIVE, capability.CAP_SYS_ADMIN) {
+		return false, nil
+	}
+
+	// if the host does not use systemd, we cannot use this workaround
+	_, err = os.Stat("/host/run/systemd/private")
+	if err != nil {
+		return false, nil
+	}
```

Comment:
With this change, when `ig` is running without CAP_SYS_ADMIN on a Linux distro without systemd, the error message will be about the non-existing systemd socket instead of being about failing later on because of the lack of CAP_SYS_ADMIN. Is it really what we want? The lack of systemd on a Linux distro without systemd is not something that the user can fix, whereas the lack of CAP_SYS_ADMIN can be fixed by the user (by creating the pod manually instead of using "kubectl debug node").

Reply:
This function is called `autoSdUnitRestart` and the flag is called `--auto-sd-unit-restart`: it was not meant to unconditionally execute the workaround but first checks if it is useful and possible. If it finds out it is not necessary (e.g. we have CAP_SYS_ADMIN) or not possible (non-systemd distro), then it does not try the workaround and continue on the normal path (and most likely fail, unless it is a command such as `ig version`, `ig --help`, etc.).

I could implement your suggestion to fail on a non-systemd distro, but then I should rename the flag to `--restart-in-sd-unit` to imply to the user that it is not conditional.

I am fine with either choice, I just want to make the flag name match its semantic.

If we later want to support non-systemd distro with Docker, we could implement another workaround that creates a privileged Docker container by talking to the dockerd socket. With my semantic, we could have two flags `--auto-sd-unit-restart` and `--auto-docker-restart`, the user could specify **both**, and ig would select whatever it finds useful and possible. With your semantic, we could have two flags `--restart-in-sd-unit` and `--restart-in-docker`, and the user must use **one of them** because using the wrong one would fail.

Reply:
> On one hand, if we almost sure it will fail, then better to not run ig at all and inform the user of why it will fail (lack of CAP_SYS_ADMIN or non systemd distro).
> On another hand, the above would prevent users running "simple" command, like ig --help.

The workarounds are now done in the local runtime, so simple commands like "version" and "help" should run without attempting the workaround.

> Moreover, if it fails due to lack of CAP_SYS_ADMIN and/or systemd, what ig would print?

At the moment, because of lack of capabilities:
```
$ kubectl debug node/minikube-docker -ti --image=albantest.azurecr.io/ig:dev3 -- ig trace exec
Creating debugging pod node-debugger-minikube-docker-khj6w with container debugger on node minikube-docker.
Error: initializing operators: initializing operator "LocalManager": initializing manager: failed to set memlock rlimit: operation not permitted
```

> If the error message explains well the problem, then let's keep your solution as is, otherwise adding explanation somewhere should help.

Ok, I'll try to update this.

> At one moment, you told this workaround could maybe not be needed in the future due to some modifications upstream in kubectl debug.
> What is the status of this modification? Could we help bring it?
> In we work upstream, we would no more need this workaround and the whole situation would be easier for everyone.

[KEP-1441](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cli/1441-kubectl-debug#profile-general) defines security profiles and https://github.com/kubernetes/kubernetes/pull/114280 implemented some of them in kubectl v1.27. Unfortunately, even in the last version, the security profile "sysadmin" with CAP_SYS_ADMIN is not yet implemented. There are still some PRs in progress about KEP-1441 such as https://github.com/kubernetes/kubernetes/pull/118647 so I assume the work in KEP-1441 with the "sysadmin" will be completed at some point, but it might take some time. I will update the documentation to reflect that.